File: 프로그래머스/2/87946. 피로도/피로도.py
def dfs(dungeons, k, cnt):
    global used
    max_cnt = cnt

    for i in range(len(dungeons)):
        if used[i] == 0 and k >= dungeons[i][0]:
            used[i] = 1
            result = dfs(dungeons, k - dungeons[i][1], cnt + 1)
            max_cnt = max(max_cnt, result)
            used[i] = 0
    
    return max_cnt

def solution(k, dungeons):
    global used
    used = [0] * len(dungeons)
    return dfs(dungeons, k, 0)

File: 프로그래머스/2/87946. 피로도/피로도.js
function dfs(dungeons, k, cnt) {
    if (k < 0) {
        li.push(cnt - 1)
        console.log(li)
        return 
    }
    for (let i = 0; i < dungeons.length; i ++) {
        if (used[i] === 0 && k >= dungeons[i][0]) {
            used[i] = 1
            dfs(dungeons, k-dungeons[i][1], cnt+1)
            used[i] = 0
        }
    }
    li.push(cnt)
    return li
}

function solution(k, dungeons) {
    let cnt = 0
    li = Array()
    used = Array.from({length: dungeons.length}, () => 0)
    return Math.max(dfs(dungeons, k, cnt))
}